src/core/ensemble_scheduler.cc

```diff
@@ -0,0 +1,531 @@
+    const std::shared_ptr<ModelInferStats>& stats,
+    const std::shared_ptr<InferRequestProvider>& request_provider,
+    const std::shared_ptr<InferResponseProvider>& response_provider,
+    std::function<void(Status)> OnComplete)
+    : is_(is), info_(info), inflight_request_counter_(0),
+      tensor_data_(info_->tensor_to_step_.size()), stats_(stats),
+      request_provider_(request_provider),
+      response_provider_(response_provider), OnComplete_(OnComplete)
+{
+  // Obtain backend handles of all models in ensemble request such that
+  // they have the same lifetime as the ensemble request to avoid unloading
+  // while the ensemble is executing.
+  for (const auto& step_info : info_->steps_) {
+    auto it = handles_.find(step_info.model_name_);
+    if (it == handles_.end()) {
+      auto it =
+          handles_
+              .emplace(std::make_pair(
+                  step_info.model_name_,
+                  std::make_shared<InferenceServer::InferBackendHandle>()))
+              .first;
+      ensemble_status_ = is_->CreateBackendHandle(
+          step_info.model_name_, -1 /* model version */, it->second);
+      if (!ensemble_status_.IsOk()) {
+        break;
+      }
+    }
+  }
+
+  if (ensemble_status_.IsOk()) {
+    const auto& request_header = request_provider_->RequestHeader();
+
+    batch_size_ = request_header.batch_size();
+    correlation_id_ = request_header.correlation_id();
+
+    for (const auto& input : request_header.input()) {
+      auto it = info_->ensemble_input_to_tensor_.find(input.name());
+      if (it != info_->ensemble_input_to_tensor_.end()) {
+        auto& tensor_data = tensor_data_[it->second];
+        tensor_data.first = input;
+        request_provider_->GetSystemMemory(it->first, &(tensor_data.second));
+      } else {
+        ensemble_status_ = Status(
+            RequestStatusCode::INVALID_ARG,
+            "unexpected input '" + input.name() +
+                "' in request header that does not map to any ensemble inputs");
+      }
+    }
+  }
+}
+
+void
+EnsembleContext::Proceed(
+    const std::shared_ptr<EnsembleContext>& context,
+    const std::shared_ptr<Step>& completed_step)
+{
+  StepList ready_steps;
+  Status status = context->PrepareSteps(completed_step, ready_steps);
+  if (status.IsOk()) {
+    ScheduleSteps(context, ready_steps);
+  }
+}
+
+Status
+EnsembleContext::PrepareSteps(
+    const std::shared_ptr<Step>& completed_step, StepList& ready_steps)
+{
+  {
+    std::lock_guard<std::mutex> lock(mutex_);
+
+    StepList res;
+    if (ensemble_status_.IsOk()) {
+      std::vector<size_t> updated_tensors;
+      ensemble_status_ = UpdateEnsembleState(completed_step, updated_tensors);
+      if (ensemble_status_.IsOk()) {
+        ensemble_status_ = GetNextSteps(updated_tensors, res);
+      }
+    }
+    // Error or no more progress (completed or deadlock)
+    if ((!ensemble_status_.IsOk()) || (inflight_request_counter_ == 0)) {
```

Comment:
inflight_steps_counter_ would be a more meaningful name. I assume this is just to be extra safe since you verify that the emsemble graph is fully connected and has no loops?

Also, shouldn't you set some kind of error if you do finish the ensemble because of a deadlock?